{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "import os\n",
    "# os.environ['CURL_CA_BUNDLE'] = ''\n",
    "# os.environ['CA_CERTS'] = '/Users/zhou/Dev_Work/ask_pdf/http_ca.crt'\n",
    "from langchain.document_loaders import PyMuPDFLoader\n",
    "from langchain import ElasticVectorSearch\n",
    "from langchain.embeddings import OpenAIEmbeddings\n",
    "from langchain.callbacks import get_openai_callback\n",
    "import yaml\n",
    "\n",
    "with open(\"./config.yml\", \"r\") as ymlfile:\n",
    "    cfg = yaml.safe_load(ymlfile)\n",
    "\n",
    "OPENAI_API_KEY = cfg[\"openai\"][\"OPENAI_API_KEY\"]\n",
    "ELASTICSEARCH_URL = cfg[\"es\"][\"elasticsearch_url\"]\n",
    "embeddings = OpenAIEmbeddings(openai_api_key=OPENAI_API_KEY)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.chains.question_answering import load_qa_chain\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "llm = ChatOpenAI(temperature=0.7, openai_api_key=OPENAI_API_KEY)\n",
    "# ?, return_source_documents=True\n",
    "chain = load_qa_chain(llm, chain_type=\"stuff\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "docsearch = ElasticVectorSearch(\n",
    "    elasticsearch_url=ELASTICSEARCH_URL,\n",
    "    index_name='company',\n",
    "    embedding=embeddings,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(page_content='The primary antibodies used included: rabbit anti-HDAC1 (Abclonal, China), mouse anti-Ac-lysine (Santa Cruz, USA), rabbit anti-acetyl-H3-K27 (Abclonal, China), rabbit anti-acetyl-H3-K9 (Abclonal, China), mouse anti-histone H3 (Beyotime), mouse anti-β-actin (Beyotime), rabbit anti-ISG15 (Beyotime), rabbit anti-Sp1 (Abclonal, China), rabbit anti-Flag tag (DYKDDDDK; CST), mouse anti-HA tag (CST), mouse anti-PEDV N monoclonal antibody (stored in our laboratory) ( 56 ).\\n\\nThe following primary antibodies were used: anti-ZO-1 (1:1000, #5406, Cell Signaling Technology, Danvers, MA, USA), anti-claudin-5 (1:1000, A10207, ABclonal, Wuhan, China), anti-NLRP3 (1:1000, A14223, ABclonal), anti-caspase-1 (1:1000, A0964, ABclonal), anti-IL-1β (1:1000, A16288, ABclonal), anti-Gasdermin D (GSDMD, #96458, 1:1000, Cell Signaling Technology), anti-SLC5A1 (1:1000, #5042, Cell Signaling Technology) and anti-SLC5A2 (1:1000, #14210, Cell Signaling Technology).', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/x.txt'}),\n",
       " Document(page_content='The anti-GAPDH antibody (Proteintech), anti-ZIKV E protein antibody (BioFront), anti-ERK1/2 antibody (ABclonal), and anti-phospho-ERK1/2 (ABclonal) were used as primary antibodies.\\n\\nThe following antibodies were used in this study: NAT8 (ABclonal), NAT8B (Bioss), SLC33A1 (Bioss), β-actin (ABclonal, Woburn, MA), GFP (Thermo Fisher scientific), and Flag (Origene) antibodies.\\n\\nSecondary antibodies including goat anti-mouse HRP-conjugated IgG (1:5,000, AS003, ABclonal) and goat anti-rabbit HRP-conjugated IgG (1:5,000, AS014, ABclonal) antibodies were used.\\n\\nRabbit polyclonal antibodies were generated by Abclonal (Cambridge, MA).\\n\\nPrimary antibodies used for Immunohistochemistry: anti-LIF (ABclonal, A1288).\\n\\nThe flg22 peptide (TRLSSGLKINSAKDDAAGLQIA) was purchased from\\xa0Abclonal\\xa0(USA).\\n\\nThe HBV1.04 (3318 bp) and HBV1.2 (3718 bp) DNA fragments were amplified and inserted into AAV vector by ABclonal MultiF Seamless Assembly Mix (RK21020; ABclonal, Wuhan, China).', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/x.txt'}),\n",
       " Document(page_content='The primary antibodies were used as follows: anti-CD36 (ABclonal, Wuhan, China, A5792, 1:1,000), anti-GAPDH (ABclonal, Wuhan, China, AC033, 1:1,000), anti-FTO (ABclonal, Wuhan, China, A1438, 1:1,000), anti-P65-S536 (ABclonal, Wuhan, China, AP0475, 1:1,000), anti-IKB-α (Proteintech, Wuhan, China, 10268-1-AP, 1: 1,000), and anti-β-actin (ABclonal, Wuhan, China, AC026, 1: 1,000).\\n\\nAntibodies targeting the following proteins were used in this study: NLRP3 (A12694, Abclonal), Trx (A01219-1, BOSTER), TXNIP (sc-166234, Santa Cruz; A9342, Abclonal), IL-1β (A1112, Abclonal), caspase-1 (BM4291, BOSTER), GSDMD (A10164, Abclonal), β-actin (60008-1-Ig, Proteintech), NOX4 (14347-1-AP, Proteintech).', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/x.txt'}),\n",
       " Document(page_content='Antibodies were as follows: c-MET Rabbit mAb (ABclonal, #A17366, 1\\u2009:\\u20091000), SLC7A11/xCT Rabbit mAb (ABclonal, #A2413, 1\\u2009:\\u20091000), GPX4 Monoclonal antibody (Proteintech, #14432-1-AP, 1\\u2009:\\u20091000), and Anti-beta Tubulin antibody (Abcam, #ab6046, 1\\u2009:\\u20091000).\\n\\nAfter three washes with PBS (Gibco, Thermo Fisher Scientific), the cells were respectively stained with Cy3-conjugated goat anti-rabbit IgG (Abclonal, Wuhan, China) or FITC-conjugated goat anti-mouse IgG (Abclonal) for 30 min at 37 °C.\\n\\nAll these constructs were generated using 2×MultiF Seamless Assembly Mix (ABclonal) and sequence verified.\\n\\nAnti-H3K4me1 (A2355), anti-H3K4me2 (A2356), and anti-H3K4me3 (A2357) were purchased from ABclonal Technology.\\n\\nThe ACTB was detected with a rabbit monoclonal antibody against human ACTB (ABclonal Technology, AC026).\\n\\nAntibodies against GOLGA2, CANX, CD63 and CD81 were purchased from ABclonal Biotechnology Co.,Ltd (China).', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/x.txt'})]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"有哪些抗体\"\n",
    "docs = docsearch.similarity_search(query=query, include_metadata=True)\n",
    "docs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. rabbit anti-HDAC1\n",
      "2. mouse anti-Ac-lysine\n",
      "3. rabbit anti-acetyl-H3-K27\n",
      "4. rabbit anti-acetyl-H3-K9\n",
      "5. mouse anti-histone H3\n",
      "6. mouse anti-β-actin\n",
      "7. rabbit anti-ISG15\n",
      "8. rabbit anti-Sp1\n",
      "9. rabbit anti-Flag tag\n",
      "10. mouse anti-HA tag\n",
      "11. anti-ZO-1\n",
      "12. anti-claudin-5\n",
      "13. anti-NLRP3\n",
      "14. anti-caspase-1\n",
      "15. anti-IL-1β\n",
      "16. anti-Gasdermin D (GSDMD)\n",
      "17. anti-SLC5A1\n",
      "18. anti-SLC5A2\n",
      "19. NAT8\n",
      "20. NAT8B\n",
      "21. SLC33A1\n",
      "22. GFP\n",
      "23. Flag\n",
      "24. anti-LIF\n",
      "25. Anti-CD36\n",
      "26. Anti-GAPDH\n",
      "27. Anti-FTO\n",
      "28. Anti-P65-S536\n",
      "29. Anti-IKB-α\n",
      "30. NLRP3\n",
      "31. Trx\n",
      "32. TXNIP\n",
      "33. NOX4\n",
      "34. c-MET Rabbit mAb\n",
      "35. SLC7A11/xCT Rabbit mAb\n",
      "36. GPX4 Monoclonal antibody\n",
      "37. Anti-beta Tubulin antibody\n",
      "38. anti-H3K4me1\n",
      "39. anti-H3K4me2\n",
      "40. anti-H3K4me3\n",
      "41. rabbit monoclonal antibody against human ACTB\n",
      "42. Antibodies against GOLGA2\n",
      "43. Antibodies against CANX\n",
      "44. Antibodies against CD63\n",
      "45. Antibodies against CD81\n",
      "Total Tokens: 1689\n",
      "Prompt Tokens: 1313\n",
      "Completion Tokens: 376\n",
      "Total Cost (USD): $0.003378\n"
     ]
    }
   ],
   "source": [
    "with get_openai_callback() as cb:\n",
    "    res = chain.run(input_documents=docs, question=query)\n",
    "    print(res)\n",
    "    print(f\"Total Tokens: {cb.total_tokens}\")\n",
    "    print(f\"Prompt Tokens: {cb.prompt_tokens}\")\n",
    "    print(f\"Completion Tokens: {cb.completion_tokens}\")\n",
    "    print(f\"Total Cost (USD): ${cb.total_cost}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- 总结 -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(page_content='Molecular Cell, Volume 65\\nSupplemental Information\\nComparative Analysis\\nof Single-Cell RNA Sequencing Methods\\nChristoph Ziegenhain, Beate Vieth, Swati Parekh, Björn Reinius, Amy Guillaumet-\\nAdkins, Martha Smets, Heinrich Leonhardt, Holger Heyn, Ines Hellmann, and Wolfgang\\nEnard', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'file_path': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'page_number': 19, 'total_pages': 33, 'format': 'PDF 1.7', 'title': 'Comparative Analysis of Single-Cell RNA Sequencing Methods', 'author': 'Christoph Ziegenhain', 'subject': 'Molecular Cell, 65 (2017) 631-647. doi:10.1016/j.molcel.2017.01.023', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170211005808+05'30'\", 'modDate': \"D:20170211005907+05'30'\", 'trapped': ''}),\n",
       " Document(page_content='Article\\nComparative Analysis of Single-Cell RNA\\nSequencing Methods\\nGraphical Abstract\\nHighlights\\nd The study represents the most comprehensive comparison of\\nscRNA-seq protocols\\nd Power simulations quantify the effect of sensitivity and\\nprecision on cost efﬁciency\\nd The study offers an informed choice among six prominent\\nscRNA-seq methods\\nd The study provides a framework for benchmarking future\\nprotocol improvements\\nAuthors\\nChristoph Ziegenhain, Beate Vieth,\\nSwati Parekh, ..., Holger Heyn,\\nInes Hellmann, Wolfgang Enard\\nCorrespondence\\nenard@bio.lmu.de\\nIn Brief\\nZiegenhain et al. generated data from\\nmouse ESCs to systematically evaluate\\nsix prominent scRNA-seq methods. They\\nused power simulations to compare cost\\nefﬁciencies, allowing for informed choice\\namong existing protocols and providing a\\nframework for future comparisons.\\nZiegenhain et al., 2017, Molecular Cell 65, 631–643\\nFebruary 16, 2017 ª 2017 Elsevier Inc.\\nhttp://dx.doi.org/10.1016/j.molcel.2017.01.023', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'file_path': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'page_number': 1, 'total_pages': 33, 'format': 'PDF 1.7', 'title': 'Comparative Analysis of Single-Cell RNA Sequencing Methods', 'author': 'Christoph Ziegenhain', 'subject': 'Molecular Cell, 65 (2017) 631-647. doi:10.1016/j.molcel.2017.01.023', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170211005808+05'30'\", 'modDate': \"D:20170211005907+05'30'\", 'trapped': ''}),\n",
       " Document(page_content='(adjusted R2 of linear regression model) across methods. Each dot represents\\na cell/bead and each box represents the median and ﬁrst and third quartiles.\\nSee also Figure S5.\\n636\\nMolecular Cell 65, 631–643, February 16, 2017', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'file_path': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'page_number': 7, 'total_pages': 33, 'format': 'PDF 1.7', 'title': 'Comparative Analysis of Single-Cell RNA Sequencing Methods', 'author': 'Christoph Ziegenhain', 'subject': 'Molecular Cell, 65 (2017) 631-647. doi:10.1016/j.molcel.2017.01.023', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170211005808+05'30'\", 'modDate': \"D:20170211005907+05'30'\", 'trapped': ''}),\n",
       " Document(page_content='Molecular Cell\\nArticle\\nComparative Analysis\\nof Single-Cell RNA Sequencing Methods\\nChristoph Ziegenhain,1 Beate Vieth,1 Swati Parekh,1 Bjo¨ rn Reinius,2,3 Amy Guillaumet-Adkins,4,5 Martha Smets,6\\nHeinrich Leonhardt,6 Holger Heyn,4,5 Ines Hellmann,1 and Wolfgang Enard1,7,*\\n1Anthropology & Human Genomics, Department of Biology II, Ludwig-Maximilians University, Großhaderner Straße 2,\\n82152 Martinsried, Germany\\n2Ludwig Institute for Cancer Research, Box 240, 171 77 Stockholm, Sweden\\n3Department of Cell and Molecular Biology, Karolinska Institutet, 171 77 Stockholm, Sweden\\n4CNAG-CRG, Centre for Genomic Regulation (CRG), Barcelona Institute of Science and Technology (BIST), 08028 Barcelona, Spain\\n5Universitat Pompeu Fabra (UPF), 08002 Barcelona, Spain\\n6Department of Biology II and Center for Integrated Protein Science Munich (CIPSM), Ludwig-Maximilians University, Großhaderner Straße 2,\\n82152 Martinsried, Germany\\n7Lead Contact\\n*Correspondence: enard@bio.lmu.de', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'file_path': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'page_number': 2, 'total_pages': 33, 'format': 'PDF 1.7', 'title': 'Comparative Analysis of Single-Cell RNA Sequencing Methods', 'author': 'Christoph Ziegenhain', 'subject': 'Molecular Cell, 65 (2017) 631-647. doi:10.1016/j.molcel.2017.01.023', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170211005808+05'30'\", 'modDate': \"D:20170211005907+05'30'\", 'trapped': ''})]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"The Most Representative Paragraph.\"\n",
    "docs1 = docsearch.similarity_search(query, include_metadata=True)\n",
    "docs1[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The article presents a comparative analysis of six prominent single-cell RNA sequencing methods, using data from mouse ESCs to evaluate their sensitivity and precision, and power simulations to compare their cost efficiency. The study offers an informed choice among existing protocols and provides a framework for future comparisons and benchmarking.'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Tokens: 770\n",
      "Prompt Tokens: 713\n",
      "Completion Tokens: 57\n",
      "Total Cost (USD): $0.00154\n"
     ]
    }
   ],
   "source": [
    "from langchain.chains.summarize import load_summarize_chain\n",
    "\n",
    "chain = load_summarize_chain(llm, chain_type=\"stuff\")\n",
    "with get_openai_callback() as cb:\n",
    "  display(chain.run(docs1))\n",
    "  print(f\"Total Tokens: {cb.total_tokens}\")\n",
    "  print(f\"Prompt Tokens: {cb.prompt_tokens}\")\n",
    "  print(f\"Completion Tokens: {cb.completion_tokens}\")\n",
    "  print(f\"Total Cost (USD): ${cb.total_cost}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- 指定路径问答 -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Any, Dict, Iterable, List, Optional, Tuple\n",
    "from langchain.docstore.document import Document\n",
    "\n",
    "\n",
    "def _default_script_query(query_vector: List[float], filter: Optional[dict]) -> Dict:\n",
    "    if filter is None:\n",
    "        filter = {\"match_all\": {}}\n",
    "    else:\n",
    "        (key, value), = filter.items()\n",
    "        filter = {\"match\": {f\"metadata.{key}.keyword\": f\"{value}\"}}\n",
    "    return {\n",
    "        \"script_score\": {\n",
    "            \"query\": filter,\n",
    "            \"script\": {\n",
    "                \"source\": \"cosineSimilarity(params.query_vector, 'vector') + 1.0\",\n",
    "                \"params\": {\"query_vector\": query_vector},\n",
    "            },\n",
    "        }\n",
    "    }\n",
    "\n",
    "\n",
    "def similarity_search_with_score(\n",
    "    self,\n",
    "    query: str,\n",
    "    k: int = 4,\n",
    "    filter: Optional[dict] = None,\n",
    "    **kwargs: Any\n",
    ") -> List[Tuple[Document, float]]:\n",
    "    \"\"\"Return docs most similar to query.\n",
    "    Args:\n",
    "        query: Text to look up documents similar to.\n",
    "        k: Number of Documents to return. Defaults to 4.\n",
    "    Returns:\n",
    "        List of Documents most similar to the query.\n",
    "    \"\"\"\n",
    "    embedding = self.embedding.embed_query(query)\n",
    "    filter = kwargs.get(\"filter\")\n",
    "    script_query = _default_script_query(embedding, filter)\n",
    "    response = self.client.search(\n",
    "        index=self.index_name, query=script_query, size=k)\n",
    "    hits = [hit for hit in response[\"hits\"][\"hits\"]]\n",
    "    # documents = [\n",
    "    #     Document(page_content=hit[\"text\"], metadata=hit[\"metadata\"]) for hit in hits\n",
    "    # ]\n",
    "    docs_and_scores = [\n",
    "        (\n",
    "            Document(page_content=hit[\"_source\"][\"text\"],\n",
    "                     metadata=hit[\"_source\"][\"metadata\"]),\n",
    "            hit['_score']\n",
    "        ) for hit in hits\n",
    "    ]\n",
    "    return docs_and_scores\n",
    "\n",
    "\n",
    "ElasticVectorSearch.similarity_search_with_score = similarity_search_with_score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(Document(page_content='Molecular Cell, Volume 65\\nSupplemental Information\\nComparative Analysis\\nof Single-Cell RNA Sequencing Methods\\nChristoph Ziegenhain, Beate Vieth, Swati Parekh, Björn Reinius, Amy Guillaumet-\\nAdkins, Martha Smets, Heinrich Leonhardt, Holger Heyn, Ines Hellmann, and Wolfgang\\nEnard', metadata={'source': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'file_path': '/Users/zhou/Dev_Work/ask_pdf/mmc2.pdf', 'page_number': 19, 'total_pages': 33, 'format': 'PDF 1.7', 'title': 'Comparative Analysis of Single-Cell RNA Sequencing Methods', 'author': 'Christoph Ziegenhain', 'subject': 'Molecular Cell, 65 (2017) 631-647. doi:10.1016/j.molcel.2017.01.023', 'keywords': '', 'creator': 'Elsevier', 'producer': 'Acrobat Distiller 8.1.0 (Windows)', 'creationDate': \"D:20170211005808+05'30'\", 'modDate': \"D:20170211005907+05'30'\", 'trapped': ''}),\n",
       " 1.7538588)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"The Most Representative Paragraph.\"\n",
    "docs2 = docsearch.similarity_search_with_score(query, filter={\"title\": \"Comparative Analysis of Single-Cell RNA Sequencing Methods\"})\n",
    "docs2[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The most representative paragraph is the \"In Brief\" section, which summarizes the key points of the study. It states that the authors evaluated six prominent scRNA-seq methods using data from mouse ESCs, and used power simulations to compare cost efficiencies. The study provides an informed choice among existing protocols and a framework for future comparisons.\n",
      "Total Tokens: 815\n",
      "Prompt Tokens: 749\n",
      "Completion Tokens: 66\n",
      "Total Cost (USD): $0.00163\n"
     ]
    }
   ],
   "source": [
    "with get_openai_callback() as cb:\n",
    "    res = chain.run(input_documents=[d[0] for d in docs2], question=query)\n",
    "    print(res)\n",
    "    print(f\"Total Tokens: {cb.total_tokens}\")\n",
    "    print(f\"Prompt Tokens: {cb.prompt_tokens}\")\n",
    "    print(f\"Completion Tokens: {cb.completion_tokens}\")\n",
    "    print(f\"Total Cost (USD): ${cb.total_cost}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "基因富集分析-DLi-HjL2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
